tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc

```diff
@@ -0,0 +1,721 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifdef GOOGLE_CUDA&& GOOGLE_TENSORRT
+#include <iterator>
+#include <limits>
+#include <memory>
+
+#include "tensorflow/compiler/tf2tensorrt/common/utils.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/convert_nodes.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/op_converter.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/op_converter_registry.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/ops/layer_utils.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/utils.h"
+#include "tensorflow/core/kernels/linalg/einsum_op_impl.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "third_party/tensorrt/NvInfer.h"
+
+#if IS_TRT_VERSION_GE(7, 1, 3, 0)
+
+namespace tensorflow {
+namespace tensorrt {
+namespace convert {
+
+namespace {
+// Finds the indices of elements in [begin, end) in array
+// [array_begin, array_end), and appends the indices to permute. This is used to
+// construct the permutation sequence for the operand with input labels
+// [array_begin, array_end) to the desired permuted labels [begin, end).
+template <typename Iterator1, typename Iterator2>
+Status FindIndices(Iterator1 begin, Iterator1 end, Iterator2 array_begin,
+                   Iterator2 array_end, std::vector<int>* permute) {
+  const int n = array_end - array_begin;
+  if (n < end - begin) {
+    return errors::Internal("Incorrect array size");
+  }
+  for (auto i = begin; i < end; i++) {
+    int idx = std::find(array_begin, array_end, *i) - array_begin;
+    if (idx >= n) {
+      return errors::Internal("Label not found");
+    }
+    permute->push_back(idx);
+  }
+  return Status::OK();
+}
+
+// Layout of the einsum dimensions: Batch, Free and Contraction indices.
+// Example: abcd,adef -> abde. The first tensor has layout BFC, the second BCF.
```

Comment:
I understand this is from the old code, but when I read this again, I don't feel it is correct. Can you double check?

here batch: ad, free:be, contract:cf. right?
may we need to change the expression to adbc,adfe -> abde
to make the remaining part of the sentence correct?

Can you explain if I am not right here?

Reply:
Right I think the comment is not correct, will fix

Reply:
applied